src/http/http-client.js

```diff
@@ -0,0 +1,218 @@
+/*
+Copyright 2020 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import { timedRequest, createCancelToken } from '../http-utils';
+import ErrorCodes from '../error-codes';
+import UploadOptionsBase from '../upload-options-base';
+import UploadError from '../upload-error';
+import HttpResponse from './http-response';
+import { v4 as uuid } from 'uuid';
+
+/**
+ * Invoked when there is an immediate retry error. Handles special cases and adds the
+ * error to the given result.
+ * @param {Error|string} e The error that occurred.
+ * @param {HttpResult} httpResult Result being provided for the current operation.
+ */
+function handleRetryError(e, httpResult) {
+    if (e && e.message && e.message === ErrorCodes.USER_CANCELLED) {
+        throw new UploadError('user cancelled the operation', ErrorCodes.USER_CANCELLED);
+    }
+    if (e && e.isAxiosError) {
+        const { response = {} } = e;
+        const { status } = response;
+
+        // only retry 5xx errors and errors that don't have a status code (which
+        // indicates some kind of network or I/O error)
+        if (status && (status < 500 || status > 599)) {
+            throw e;
```

Comment:
Great callout, thanks for pointing it out. In the case of a network error, `status` will be undefined so it will be handled by the `if` statement above. I double-checked the unit tests and saw that I didn't have a case to verify this, so I added one and it passed successfully. In short: network errors will be retried.